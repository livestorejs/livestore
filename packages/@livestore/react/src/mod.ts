export { LiveStoreContext, type ReactApi } from './LiveStoreContext.js'
export { useStore, withReactApi } from './useStore.js'
export { LiveStoreProvider } from './LiveStoreProvider.js'
export { useStackInfo } from './utils/stack-info.js'
export { useQuery, useQueryRef } from './useQuery.js'
export {
  useClientDocument,
  type StateSetters,
  type SetStateAction,
  type Dispatch,
  type UseRowResult as UseStateResult,
} from './useClientDocument.js'
