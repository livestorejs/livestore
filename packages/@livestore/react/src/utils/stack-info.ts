import { extractStackInfoFromStackTrace, type StackInfo } from '@livestore/livestore'
import React from 'react'

export const originalStackLimit = Error.stackTraceLimit

export const useStackInfo = (): StackInfo =>
  React.useMemo(() => {
    Error.stackTraceLimit = 10
    // eslint-disable-next-line unicorn/error-message
    const stack = new Error().stack!
    Error.stackTraceLimit = originalStackLimit
    return extractStackInfoFromStackTrace(stack)
  }, [])
