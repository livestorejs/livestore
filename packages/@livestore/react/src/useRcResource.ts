import * as React from 'react'

/**
 * Creates a reference-counted resource object that is "stable" across React lifecycles.
 *
 * The hook is primarily intended for creating stateful objects or entities where:
 * 1) reference identity is crucial (e.g. stateful objects, references, ...)
 *    and needed to persist across multiple component instances
 * 2) the `create`/`dispose` functions might be effectful and can be called exactly once per key
 *
 * **Goals:**
 * - Guarantee stable reference identity for stateful entities that are created via the `create` function.
 * - Strong guarantees about the `create`/`dispose` calls (i.e. exactly one `create`/`dispose` call per key)
 * - Ensure that such state is retained across component re-mounts, even in development environments like
 *   React Strict Mode or during Fast Refresh.
 * - Automatically handle resource disposal by decrementing a reference count and disposing of the resource
 *   when no components are using it.
 *
 * **Behavior:**
 * - On the first render with a specific key, the `create` function is invoked to create the stateful entity,
 *   and the resource is stored in a cache with a reference count set to 1.
 * - If another component renders with the same key, the cached entity is reused and its reference count is incremented.
 * - When a component renders with a new key, the previous key's reference count is decremented and, if it reaches zero,
 *   the `dispose` function is called for that resource.
 * - Upon component unmount, the reference count is decremented, leading to disposal (via the `dispose` function)
 *   if the reference count drops to zero. An unmount is either detected via React's `useEffect` callback or
 *   in the useMemo hook when the key changes.
 * 
 * Why this is needed in LiveStore:
 * Let's first take a look at the "trivial implementation":
 * 
 * ```ts
 * const useSimpleResource = <T>(create: () => T, dispose: (resource: T) => void) => {
 *     const val = React.useMemo(() => create(), [create])
 *
 *     React.useEffect(() => {
 *       return () => {
 *         dispose(val)
 *       }
 *     }, [dispose, val])

 *     return val
 * }
 * ```
 * 
 * LiveStore uses this hook to create LiveQuery instances which are stateful and must not be leaked.
 * The simple implementation above would leak the LiveQuery instance if the component is unmounted or props change.
 *
 * **Usage:**
 * ```tsx
 * // Create a stateful object instance for a unique key and share it between components.
 * const statefulObject = useRcResource(
 *   'stable-object-key',
 *   () => createObjectInstance(),
 *   (object) => object.dispose()
 * );
 * ```
 *
 * **Caveats:**
 * - The `create` function is intentionally omitted from the dependency array in `useMemo` to prevent
 *   unintended re-creations of the stateful entity. Avoid closing over changing values within `create`
 *   or include them in the `key`.
 * - Ensure that the `dispose` function is stable or properly memoized as its reference is used in a `useEffect`.
 * - Although the caching mechanism prevents duplicate instance creation for the same key, the strategy
 *   can interact in unexpected ways with React’s development patterns. Please report any issues if encountered.
 *
 * @template T The type of the stateful entity managed by the hook.
 * @param key A unique identifier for the stateful entity. A change in this key triggers a disposal of the previous resource.
 * @param create Function to create the stateful entity when it does not exist in the cache.
 * @param dispose Function to dispose of the stateful entity when it’s no longer needed. Needs to be stable.
 * @param _options Optional. Additional options such as a debug print callback for logging purposes.
 * @returns The stateful entity corresponding to the provided key.
 */
export const useRcResource = <T>(
  key: string,
  create: () => T,
  dispose: (resource: NoInfer<T>) => void,
  _options?: { debugPrint?: (resource: NoInfer<T>) => ReadonlyArray<any> },
): T => {
  const keyRef = React.useRef<string | undefined>(undefined)
  const didDisposeInMemo = React.useRef(false)

  // biome-ignore lint/correctness/useExhaustiveDependencies: Dependency is deliberately limited to `key` to avoid unintended re-creations.
  const resource = React.useMemo(() => {
    // console.debug('useMemo', key)
    if (didDisposeInMemo.current) {
      // console.debug('useMemo', key, 'skip')
      const cachedItem = cache.get(key)
      if (cachedItem !== undefined && cachedItem._tag === 'active') {
        return cachedItem.resource
      }
    }

    // Check if the key has changed (or is undefined)
    if (keyRef.current !== undefined && keyRef.current !== key) {
      // If the key has changed, decrement the reference on the previous key
      const previousKey = keyRef.current
      const cachedItemForPreviousKey = cache.get(previousKey)
      if (cachedItemForPreviousKey !== undefined && cachedItemForPreviousKey._tag === 'active') {
        // previousKeyRef.current = previousKey
        cachedItemForPreviousKey.rc--

        // console.debug('useMemo', key, 'rc--', previousKey, cachedItemForPreviousKey.rc)

        if (cachedItemForPreviousKey.rc === 0) {
          // Clean up the stateful resource if no longer referenced
          dispose(cachedItemForPreviousKey.resource)
          cache.set(previousKey, { _tag: 'destroyed' })
          didDisposeInMemo.current = true
        }
      }
    }

    const cachedItem = cache.get(key)
    if (cachedItem !== undefined && cachedItem._tag === 'active') {
      // In React Strict Mode, the `useMemo` hook is called multiple times,
      // so we only increment the reference from the first call for this component.
      cachedItem.rc++
      // console.debug('rc++', cachedItem.rc, ...(_options?.debugPrint?.(cachedItem.resource) ?? []))

      return cachedItem.resource
    }

    // Create a new stateful resource if not cached
    const resource = create()
    cache.set(key, { _tag: 'active', rc: 1, resource })
    return resource
  }, [key])

  // biome-ignore lint/correctness/useExhaustiveDependencies: We assume the `dispose` function is stable and won't change across renders
  React.useEffect(() => {
    return () => {
      if (didDisposeInMemo.current) {
        // console.debug('unmount', keyRef.current, 'skip')
        didDisposeInMemo.current = false
        return
      }

      // console.debug('unmount', keyRef.current)
      const cachedItem = cache.get(key)
      // If the stateful resource is already cleaned up, do nothing.
      if (cachedItem === undefined || cachedItem._tag === 'destroyed') return

      cachedItem.rc--

      // console.debug('rc--', cachedItem.rc, ...(_options?.debugPrint?.(cachedItem.resource) ?? []))

      if (cachedItem.rc === 0) {
        dispose(cachedItem.resource)
        cache.delete(key)
      }
    }
  }, [key])

  keyRef.current = key

  return resource
}

// NOTE Given `useMemo` will be called multiple times (e.g. when using React Strict mode or Fast Refresh),
// we are using this cache to avoid starting multiple queries/spans for the same component.
// This is somewhat against some recommended React best practices, but it should be fine in our case below.
// Please definitely open an issue if you see or run into any problems with this approach!
const cache = new Map<
  string,
  | {
      _tag: 'active'
      rc: number
      resource: any
    }
  | {
      _tag: 'destroyed'
    }
>()

export const __resetUseRcResourceCache = () => {
  cache.clear()
}
