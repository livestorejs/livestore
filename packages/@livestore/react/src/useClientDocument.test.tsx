import * as LiveStore from '@livestore/livestore'
import { getSimplifiedRootSpan } from '@livestore/livestore/internal/testing-utils'
import { Effect, ReadonlyRecord, Schema } from '@livestore/utils/effect'
import { Vitest } from '@livestore/utils-dev/node-vitest'
import * as otel from '@opentelemetry/api'
import { BasicTracerProvider, InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import * as ReactTesting from '@testing-library/react'
import React from 'react'
import { beforeEach, expect, it } from 'vitest'

import { events, makeTodoMvcReact, tables } from './__tests__/fixture.js'
import type * as LiveStoreReact from './mod.js'
import { __resetUseRcResourceCache } from './useRcResource.js'

// const strictMode = process.env.REACT_STRICT_MODE !== undefined

// NOTE running tests concurrently doesn't work with the default global db graph
Vitest.describe('useClientDocument', () => {
  beforeEach(() => {
    __resetUseRcResourceCache()
  })

  Vitest.scopedLive('should update the data based on component key', () =>
    Effect.gen(function* () {
      const { wrapper, store, renderCount } = yield* makeTodoMvcReact({})

      const { result, rerender } = ReactTesting.renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState, id] = store.useClientDocument(tables.userInfo, userId)
          return { state, setState, id }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)
      expect(store.reactivityGraph.getSnapshot({ includeResults: true })).toMatchSnapshot()
      store.commit(tables.userInfo.set({ username: 'username_u2' }, 'u2'))

      rerender('u2')

      expect(store.reactivityGraph.getSnapshot({ includeResults: true })).toMatchSnapshot()
      expect(result.current.id).toBe('u2')
      expect(result.current.state.username).toBe('username_u2')
      expect(renderCount.val).toBe(2)
    }),
  )

  // TODO add a test that makes sure React doesn't re-render when a setter is used to set the same value

  Vitest.scopedLive('should update the data reactively - via setState', () =>
    Effect.gen(function* () {
      const { wrapper, store, renderCount } = yield* makeTodoMvcReact({})

      const { result } = ReactTesting.renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState, id] = store.useClientDocument(tables.userInfo, userId)
          return { state, setState, id }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      ReactTesting.act(() => result.current.setState({ username: 'username_u1_hello' }))

      expect(result.current.id).toBe('u1')
      expect(result.current.state.username).toBe('username_u1_hello')
      expect(renderCount.val).toBe(2)
    }),
  )

  Vitest.scopedLive('should update the data reactively - via raw store commit', () =>
    Effect.gen(function* () {
      const { wrapper, store, renderCount } = yield* makeTodoMvcReact({})

      const { result } = ReactTesting.renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState, id] = store.useClientDocument(tables.userInfo, userId)
          return { state, setState, id }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      ReactTesting.act(() => store.commit(events.UserInfoSet({ username: 'username_u1_hello' }, 'u1')))

      expect(result.current.id).toBe('u1')
      expect(result.current.state.username).toBe('username_u1_hello')
      expect(renderCount.val).toBe(2)
    }),
  )

  Vitest.scopedLive('should work for a larger app', () =>
    Effect.gen(function* () {
      const { wrapper, store, renderCount } = yield* makeTodoMvcReact({})

      const allTodos$ = LiveStore.queryDb(
        { query: `select * from todos`, schema: Schema.Array(tables.todos.rowSchema) },
        { label: 'allTodos' },
      )

      let globalSetState: LiveStoreReact.StateSetters<typeof tables.AppRouterSchema> | undefined
      const AppRouter: React.FC = () => {
        renderCount.inc()

        const [state, setState] = store.useClientDocument(tables.AppRouterSchema, 'singleton')

        globalSetState = setState

        return (
          <div>
            <TasksList setTaskId={(taskId) => setState({ currentTaskId: taskId })} />
            <div role="current-id">Current Task Id: {state.currentTaskId ?? '-'}</div>
            {state.currentTaskId ? <TaskDetails id={state.currentTaskId} /> : <div>Click on a task to see details</div>}
          </div>
        )
      }

      const TasksList: React.FC<{ setTaskId: (_: string) => void }> = ({ setTaskId }) => {
        const allTodos = store.useQuery(allTodos$)

        return (
          <div>
            {allTodos.map((_) => (
              <div key={_.id} onClick={() => setTaskId(_.id)}>
                {_.id}
              </div>
            ))}
          </div>
        )
      }

      const TaskDetails: React.FC<{ id: string }> = ({ id }) => {
        const todo = store.useQuery(LiveStore.queryDb(tables.todos.where({ id }).first(), { deps: id }))
        return <div role="content">{JSON.stringify(todo)}</div>
      }

      const renderResult = ReactTesting.render(<AppRouter />, { wrapper })

      expect(renderCount.val).toBe(1)

      ReactTesting.act(() =>
        store.commit(
          LiveStore.rawSqlEvent({
            sql: LiveStore.sql`INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)`,
          }),
        ),
      )

      expect(renderCount.val).toBe(1)
      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: -"')

      ReactTesting.act(() => globalSetState!({ currentTaskId: 't1' }))

      expect(renderCount.val).toBe(2)
      expect(renderResult.getByRole('content').innerHTML).toMatchInlineSnapshot(
        `"{"id":"t1","text":"buy milk","completed":false}"`,
      )

      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t1"')

      ReactTesting.act(() =>
        store.commit(
          events.todoCreated({ id: 't2', text: 'buy eggs', completed: false }),
          events.AppRouterSet({ currentTaskId: 't2' }),
          events.todoCreated({ id: 't3', text: 'buy bread', completed: false }),
        ),
      )

      expect(renderCount.val).toBe(3)
      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t2"')
    }),
  )

  Vitest.scopedLive('should work for a useClientDocument query chained with a useTemporary query', () =>
    Effect.gen(function* () {
      const { store, wrapper, renderCount } = yield* makeTodoMvcReact({})

      store.commit(
        events.todoCreated({ id: 't1', text: 'buy milk', completed: false }),
        events.todoCreated({ id: 't2', text: 'buy bread', completed: false }),
      )

      const { result, unmount, rerender } = ReactTesting.renderHook(
        (userId: string) => {
          renderCount.inc()

          const [_row, _setRow, _id, rowState$] = store.useClientDocument(tables.userInfo, userId)
          const todos = store.useQuery(
            LiveStore.queryDb(
              (get) => tables.todos.where('text', 'LIKE', `%${get(rowState$).text}%`),
              // TODO find a way where explicit `userId` is not needed here
              // possibly by automatically understanding the `get(rowState$)` dependency
              { label: 'todosFiltered', deps: userId },
            ),
            // TODO introduce a `deps` array which is only needed when a query is parametric
          )

          return { todos }
        },
        { wrapper, initialProps: 'u1' },
      )

      ReactTesting.act(() => store.commit(events.UserInfoSet({ username: 'username_u2', text: 'milk' }, 'u2')))

      expect(result.current.todos.length).toBe(2)
      expect(renderCount.val).toBe(1)

      rerender('u2')

      expect(result.current.todos.length).toBe(1)
      expect(renderCount.val).toBe(2)

      unmount()
    }),
  )

  Vitest.describe('otel', () => {
    it.each([{ strictMode: true }, { strictMode: false }])(
      'should update the data based on component key strictMode=%s',
      async ({ strictMode }) => {
        const exporter = new InMemorySpanExporter()

        const provider = new BasicTracerProvider({
          spanProcessors: [new SimpleSpanProcessor(exporter)],
        })

        const otelTracer = provider.getTracer(`testing-${strictMode ? 'strict' : 'non-strict'}`)

        const span = otelTracer.startSpan('test-root')
        const otelContext = otel.trace.setSpan(otel.context.active(), span)

        await Effect.gen(function* () {
          const { wrapper, store, renderCount } = yield* makeTodoMvcReact({
            otelContext,
            otelTracer,
            strictMode,
          })

          const { result, rerender, unmount } = ReactTesting.renderHook(
            (userId: string) => {
              renderCount.inc()

              const [state, setState, id] = store.useClientDocument(tables.userInfo, userId)
              return { state, setState, id }
            },
            { wrapper, initialProps: 'u1' },
          )

          expect(result.current.id).toBe('u1')
          expect(result.current.state.username).toBe('')
          expect(renderCount.val).toBe(1)

          // For u2 we'll make sure that the row already exists,
          // so the lazy `insert` will be skipped
          ReactTesting.act(() => store.commit(events.UserInfoSet({ username: 'username_u2' }, 'u2')))

          rerender('u2')

          expect(result.current.id).toBe('u2')
          expect(result.current.state.username).toBe('username_u2')
          expect(renderCount.val).toBe(2)

          unmount()
          span.end()
        }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

        await provider.forceFlush()

        const mapAttributes = (attributes: otel.Attributes) => {
          return ReadonlyRecord.map(attributes, (val, key) => {
            if (key === 'code.stacktrace') {
              return '<STACKTRACE>'
            } else if (key === 'firstStackInfo') {
              const stackInfo = JSON.parse(val as string) as LiveStore.StackInfo
              // stackInfo.frames.shift() // Removes `renderHook.wrapper` from the stack
              stackInfo.frames.forEach((_) => {
                if (_.name.includes('renderHook.wrapper')) {
                  _.name = 'renderHook.wrapper'
                }
                _.filePath = '__REPLACED_FOR_SNAPSHOT__'
              })
              return JSON.stringify(stackInfo)
            }
            return val
          })
        }

        expect(getSimplifiedRootSpan(exporter, mapAttributes)).toMatchSnapshot()

        await provider.shutdown()
      },
    )
  })
})
