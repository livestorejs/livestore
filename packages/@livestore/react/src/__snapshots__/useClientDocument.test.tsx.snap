// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useClientDocument > otel > should update the data based on component key strictMode={ strictMode: false } 1`] = `
{
  "_name": "createStore",
  "attributes": {
    "debugInstanceId": "test",
    "storeId": "default",
  },
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "client-session-sync-processor:pull",
      "attributes": {
        "code.stacktrace": "<STACKTRACE>",
        "span.label": "⚠︎ Interrupted",
        "status.interrupted": true,
      },
    },
    {
      "_name": "LiveStore:sync",
    },
    {
      "_name": "LiveStore:commits",
      "children": [
        {
          "_name": "LiveStore:commit",
          "attributes": {
            "livestore.eventTags": [
              "UserInfoSet",
            ],
            "livestore.eventsCount": 1,
          },
          "children": [
            {
              "_name": "livestore.in-memory-db:execute",
              "attributes": {
                "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(value, ?, json(?))
    ",
              },
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "LiveStore:useClientDocument:UserInfo:u1",
          "attributes": {
            "firstStackInfo": "{"frames":[{"name":"Store.store.useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
            "label": "SELECT * FROM 'UserInfo' WHERE id = ?",
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 0,
              },
            },
            {
              "_name": "LiveStore:commit",
              "attributes": {
                "livestore.commitLabel": "UserInfo.set:u1",
                "livestore.eventTags": [
                  "UserInfoSet",
                ],
                "livestore.eventsCount": 1,
              },
              "children": [
                {
                  "_name": "livestore.in-memory-db:execute",
                  "attributes": {
                    "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(json_set(value, ?, json(?)), ?, json(?))
    ",
                  },
                },
              ],
            },
            {
              "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
              "attributes": {
                "livestore.debugRefreshReason": "useQuery:initial-run:UserInfo.get:u1",
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
              "children": [
                {
                  "_name": "sql-in-memory-select",
                  "attributes": {
                    "sql.cached": false,
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                },
              ],
            },
            {
              "_name": "LiveStore.subscribe",
              "attributes": {
                "label": "UserInfo.get:u1",
                "queryLabel": "UserInfo.get:u1",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "livestore.debugRefreshReason": "commit",
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": false,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          "_name": "LiveStore:useClientDocument:UserInfo:u2",
          "attributes": {
            "firstStackInfo": "{"frames":[{"name":"Store.store.useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
            "label": "SELECT * FROM 'UserInfo' WHERE id = ?",
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
            },
            {
              "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
              "attributes": {
                "livestore.debugRefreshReason": "useQuery:initial-run:UserInfo.get:u2",
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
              "children": [
                {
                  "_name": "sql-in-memory-select",
                  "attributes": {
                    "sql.cached": false,
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                },
              ],
            },
            {
              "_name": "LiveStore.subscribe",
              "attributes": {
                "label": "UserInfo.get:u2",
                "queryLabel": "UserInfo.get:u2",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "livestore.debugRefreshReason": "subscribe-initial-run:UserInfo.get:u2",
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": true,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`useClientDocument > otel > should update the data based on component key strictMode={ strictMode: true } 1`] = `
{
  "_name": "createStore",
  "attributes": {
    "debugInstanceId": "test",
    "storeId": "default",
  },
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "client-session-sync-processor:pull",
      "attributes": {
        "code.stacktrace": "<STACKTRACE>",
        "span.label": "⚠︎ Interrupted",
        "status.interrupted": true,
      },
    },
    {
      "_name": "LiveStore:sync",
    },
    {
      "_name": "LiveStore:commits",
      "children": [
        {
          "_name": "LiveStore:commit",
          "attributes": {
            "livestore.eventTags": [
              "UserInfoSet",
            ],
            "livestore.eventsCount": 1,
          },
          "children": [
            {
              "_name": "livestore.in-memory-db:execute",
              "attributes": {
                "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(value, ?, json(?))
    ",
              },
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "LiveStore:useClientDocument:UserInfo:u1",
          "attributes": {
            "firstStackInfo": "{"frames":[{"name":"Store.store.useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
            "label": "SELECT * FROM 'UserInfo' WHERE id = ?",
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 0,
              },
            },
            {
              "_name": "LiveStore:commit",
              "attributes": {
                "livestore.commitLabel": "UserInfo.set:u1",
                "livestore.eventTags": [
                  "UserInfoSet",
                ],
                "livestore.eventsCount": 1,
              },
              "children": [
                {
                  "_name": "livestore.in-memory-db:execute",
                  "attributes": {
                    "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(json_set(value, ?, json(?)), ?, json(?))
    ",
                  },
                },
              ],
            },
            {
              "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
              "attributes": {
                "livestore.debugRefreshReason": "useQuery:initial-run:UserInfo.get:u1",
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
              "children": [
                {
                  "_name": "sql-in-memory-select",
                  "attributes": {
                    "sql.cached": false,
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                },
              ],
            },
            {
              "_name": "LiveStore.subscribe",
              "attributes": {
                "label": "UserInfo.get:u1",
                "queryLabel": "UserInfo.get:u1",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "livestore.debugRefreshReason": "commit",
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": false,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`useClientDocument > should update the data based on component key 1`] = `
{
  "atoms": [
    {
      "_tag": "ref",
      "id": "node-1",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:todos",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-2",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:app",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-3",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:UserInfo",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 1,
      "sub": [],
      "super": [
        "node-5",
      ],
    },
    {
      "_tag": "ref",
      "id": "node-4",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:AppRouter",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "thunk",
      "id": "node-5",
      "isDestroyed": false,
      "isDirty": false,
      "label": "UserInfo.get:u1:results",
      "meta": {
        "liveStoreThunkType": "db.result",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "{"username":"","text":""}",
      },
      "recomputations": 1,
      "sub": [
        "node-3",
      ],
      "super": [
        "node-6",
      ],
    },
  ],
  "deferredEffects": [],
  "effects": [
    {
      "_tag": "effect",
      "id": "node-6",
      "invocations": 1,
      "isDestroyed": false,
      "label": "subscribe:UserInfo.get:u1",
      "sub": [
        "node-5",
      ],
    },
  ],
}
`;

exports[`useClientDocument > should update the data based on component key 2`] = `
{
  "atoms": [
    {
      "_tag": "ref",
      "id": "node-1",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:todos",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-2",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:app",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-3",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:UserInfo",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 2,
      "sub": [],
      "super": [
        "node-7",
      ],
    },
    {
      "_tag": "ref",
      "id": "node-4",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:AppRouter",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "thunk",
      "id": "node-7",
      "isDestroyed": false,
      "isDirty": false,
      "label": "UserInfo.get:u2:results",
      "meta": {
        "liveStoreThunkType": "db.result",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "{"username":"username_u2","text":""}",
      },
      "recomputations": 2,
      "sub": [
        "node-3",
      ],
      "super": [
        "node-8",
      ],
    },
  ],
  "deferredEffects": [],
  "effects": [
    {
      "_tag": "effect",
      "id": "node-8",
      "invocations": 1,
      "isDestroyed": false,
      "label": "subscribe:UserInfo.get:u2",
      "sub": [
        "node-7",
      ],
    },
  ],
}
`;
